### PR TITLE
refine `SubqueryForSet`

### PR DESCRIPTION
**Pull request by SeaRise**
### What problem does this PR solve?

Issue Number: ref #4118

Problem Summary:

### What is changed and how it works?
- move `SubqueryForSet` from `ExpressionAnalyzer.h` to `SubqueryForSet.h`
- move `subqueries_for_sets` from `InterpreterDAG` to `DAGContext`

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Flash/Coprocessor/InterpreterDAG.cpp [hunk 1]
```diff
@@ -61,26 +66,23 @@ BlockIO InterpreterDAG::execute()
...
 
     /// add union to run in parallel if needed
-    if (context.getDAGContext()->isMPPTask())
+    if (dagContext().isMPPTask())
         /// MPPTask do not need the returned blocks.
         executeUnion(pipeline, max_streams, dagContext().log, /*ignore_block=*/true);
     else
         executeUnion(pipeline, max_streams, dagContext().log);
-    if (!subqueries_for_sets.empty())
+    if (dagContext().subqueries_for_sets.empty())
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/InterpreterDAG.cpp:80]:
```suggestion
    if (!dagContext().subqueries_for_sets.empty())
```

File: dbms/src/Flash/Coprocessor/InterpreterDAG.cpp [hunk 2]
```diff
@@ -61,26 +66,23 @@ BlockIO InterpreterDAG::execute()
...
 
     /// add union to run in parallel if needed
-    if (context.getDAGContext()->isMPPTask())
+    if (dagContext().isMPPTask())
         /// MPPTask do not need the returned blocks.
         executeUnion(pipeline, max_streams, dagContext().log, /*ignore_block=*/true);
     else
         executeUnion(pipeline, max_streams, dagContext().log);
-    if (!subqueries_for_sets.empty())
+    if (!dagContext().subqueries_for_sets.empty())
```
Comment (ywqzzy) [inline dbms/src/Flash/Coprocessor/InterpreterDAG.cpp:80]: How about wrap it into has_subquery()?

File: dbms/src/Flash/Coprocessor/InterpreterDAG.cpp [hunk 3]
```diff
@@ -61,26 +66,23 @@ BlockIO InterpreterDAG::execute()
...
-    std::vector<SubqueriesForSets> subqueries_for_sets;
-    BlockInputStreams streams = executeQueryBlock(*dag.getRootQueryBlock(), subqueries_for_sets);
+
+    BlockInputStreams streams = executeQueryBlock(*dag.getRootQueryBlock());
     DAGPipeline pipeline;
     pipeline.streams = streams;
 
     /// add union to run in parallel if needed
-    if (context.getDAGContext()->isMPPTask())
+    if (dagContext().isMPPTask())
```
[the remark this answers is not in the document]
Comment (SeaRise) [reply dbms/src/Flash/Coprocessor/InterpreterDAG.cpp:75]: add a function `isMppTask()` in `InterpreterDAG.cpp`?

[the remark this answers is not in the document]
Comment (SeaRise) [reply dbms/src/Flash/Coprocessor/InterpreterDAG.cpp:75]: I tend not to..

File: dbms/src/Interpreters/SubqueryForSet.h [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+#pragma once
+
+#include <common/types.h>
+
+#include <memory>
+#include <unordered_map>
+
+namespace DB
+{
+class Join;
```
Comment (SeaRise) [inline dbms/src/Interpreters/SubqueryForSet.h:24]: clear using class xxx;

**Review by ywqzzy — APPROVED** [review]

**Review by windtalker — APPROVED** [review]
LGTM

**Pull request merged**